File: homy/modules/clock/clock.py
# Clock widget — client-side only, no backend routes
WIDGETS = [
    {
        "type": "clock",
        "name": "Uhr / Clock",
        "default_size_x": 4,
        "default_size_y": 2,
        "config_schema": {
            "timezone": {
                "type": "text",
                "label": "Zeitzone (IANA, z.B. Europe/Berlin)",
                "default": "Europe/Berlin",
            },
            "format_24h": {
                "type": "select",
                "label": "24-Stunden-Format",
                "options": ["Ja", "Nein"],
                "default": "Ja",
            },
            "show_seconds": {
                "type": "select",
                "label": "Sekunden anzeigen",
                "options": ["Ja", "Nein"],
                "default": "Ja",
            },
            "show_date": {
                "type": "select",
                "label": "Datum anzeigen",
                "options": ["Ja", "Nein"],
                "default": "Ja",
            },
            "show_timezone": {
                "type": "select",
                "label": "Zeitzone anzeigen",
                "options": ["Ja", "Nein"],
                "default": "Nein",
            },
            "label": {
                "type": "text",
                "label": "Beschriftung (optional)",
                "default": "",
            },
        },
    }
]

def register(app):
    pass

File: homy/modules/clock/clock.js
window.WidgetRegistry.register('clock', {
    _timers: {},

    render(container, widgetData, config) {
        const tz = config.timezone || 'Europe/Berlin';
        const use24 = config.format_24h !== 'Nein';
        const showSec = config.show_seconds !== 'Nein';
        const showDate = config.show_date !== 'Nein';
        const showTz = config.show_timezone === 'Ja';
        const label = config.label || '';

        if (this._timers[widgetData.id]) {
            clearInterval(this._timers[widgetData.id]);
        }

        const tick = () => {
            const now = new Date();
            const timeOpts = {
                timeZone: tz,
                hour: '2-digit',
                minute: '2-digit',
                second: showSec ? '2-digit' : undefined,
                hour12: !use24,
            };
            const timeStr = now.toLocaleTimeString(window.i18n?.currentLocale || 'en-US', timeOpts);
            let dateStr = '';
            if (showDate) {
                dateStr = now.toLocaleDateString(window.i18n?.currentLocale || 'en-US', {
                    timeZone: tz,
                    weekday: 'long',
                    day: 'numeric',
                    month: 'long',
                });
            }
            container.innerHTML = `
                <div class="clock-widget">
                    ${label ? `<div class="clock-label muted-text">${window.escapeHtml(label)}</div>` : ''}
                    <div class="clock-time">${timeStr}</div>
                    ${showDate ? `<div class="clock-date">${dateStr}</div>` : ''}
                    ${showTz ? `<div class="clock-tz muted-text">${window.escapeHtml(tz)}</div>` : ''}
                </div>
            `;
        };

        tick();
        this._timers[widgetData.id] = setInterval(tick, showSec ? 1000 : 30000);
    },

    onRemove(widgetId) {
        if (this._timers[widgetId]) {
            clearInterval(this._timers[widgetId]);
            delete this._timers[widgetId];
        }
    },
});
